source/MRMesh/MRFeatures.h

```diff
@@ -130,28 +140,63 @@ namespace Primitives
 // `infiniteExtend` is how large we make "infinite" objects. Half-infinite objects divide this by 2.
 [[nodiscard]] MRMESH_API std::shared_ptr<VisualObject> primitiveToObject( const Primitives::Variant& primitive, float infiniteExtent );
 
-//! Stores the distance between two objects, and the closest points on them.
-struct DistanceResult
+//! Stores the results of measuring two objects relative to one another.
+struct MeasureResult
 {
     enum class Status
     {
         ok = 0,
         //! Algorithms set this if this when something isn't yet implemented.
-        not_implemented,
+        notImplemented,
         //! Algorithms set this when the calculation doesn't make sense for those object types.
-        not_applicable,
-        //! This is set automatically if either `distance` or at least one of the points isn't finite. But you can set this from an algorithm too.
-        not_finite,
+        //! This result can be based on object parameters, but not on their relative location.
+        badFeaturePair,
+        //! Can't be computed because of how the objects are located relative to each other.
+        badRelativeLocation,
+        //! The result was not finite. This is set automatically if you return non-finite values, but you can also set this manually.
+        notFinite,
+    };
+
+    struct BasicPart
+    {
+        Status status = Status::notImplemented;
+        [[nodiscard]] operator bool() const { return status == Status::ok; }
     };
-    Status status = Status::ok;
 
-    // This is a separate field because it can be negative.
-    float distance = 0;
+    struct Distance : BasicPart
+    {
+        // This is a separate field because it can be negative.
+        float distance = 0;
 
-    Vector3f closestPointA;
-    Vector3f closestPointB;
+        Vector3f closestPointA;
+        Vector3f closestPointB;
+
+        [[nodiscard]] Vector3f closestPointFor( bool b ) const { return b ? closestPointB : closestPointA; }
+    };
+    Distance distance;
 
-    [[nodiscard]] operator bool() const { return status == Status::ok; }
+    struct Angle : BasicPart
+    {
+        Vector3f pointA;
+        Vector3f pointB;
+        [[nodiscard]] Vector3f pointFor( bool b ) const { return b ? pointB : pointA; }
+
+        Vector3f dirA; // Normalized.
+        Vector3f dirB; // ^
+        [[nodiscard]] Vector3f dirFor( bool b ) const { return b ? dirB : dirA; }
+
+        /// Whether `dir{A,B}` is a surface normal or a line direction.
+        bool isSurfaceNormalA = false;
+        bool isSurfaceNormalB = false;
```

Comment:
whitch surface?

Reply:
also why is it needed?

Reply:
It's only needed for better visualization. If true, the object has a surface and this is the normal. If false, the object is a like and this is its direction.